packages/react-core/src/filters/viewportFeaturesBinary.js

```diff
@@ -0,0 +1,169 @@
+
+function addIntersectedFeaturesInTile({ map, data, viewportIntersection, type, uniqueIdProperty }) {
+  const indices = getIndices(data);
+  const { positions } = data;
+
+  for (let i = 0; i < indices.length - 1; i++) {
+    const startIndex = indices[i];
+    const endIndex = indices[i + 1];
+    const uniquePropertyValue = getUniquePropertyValue(data, startIndex, uniqueIdProperty, map);
+
+    if (uniquePropertyValue && !map.has(uniquePropertyValue)) {
+      const ringCoordinates = getRingCoordinatesFor(startIndex, endIndex, positions);
+      if (intersects(getFeatureByType(ringCoordinates, type), viewportIntersection)) {
+        map.set(uniquePropertyValue, parseProperties(data, startIndex));
+      }
+    }
+  }
+}
+
+function getIndices(data) {
+  const indices = data.primitivePolygonIndices || data.pathIndices || data.pointIndices;
+  return indices.value;
+}
+
+function getFeatureId(data, startIndex) {
+  return data.featureIds.value[startIndex];
+}
+
+function getPropertiesFromTile(data, startIndex) {
+  const featureId = getFeatureId(data, startIndex);
+  const { properties, numericProps } = data;
+  const result = {
+    properties: properties[featureId],
+    numericProps: {}
+  };
+
+  for (const key in numericProps) {
+    result.numericProps[key] = numericProps[key].value[startIndex];
+  }
+
+  return result;
+}
+
+function parseProperties(data, startIndex) {
+  const { properties, numericProps } = getPropertiesFromTile(data, startIndex);
+  return Object.assign({}, properties, numericProps);
+}
+
+function getUniquePropertyValue (data, startIndex, uniqueIdProperty, map) {
+  const { properties, numericProps } = getPropertiesFromTile(data, startIndex);
+  if (uniqueIdProperty) {
+    return numericProps[uniqueIdProperty] || properties[uniqueIdProperty];
+  }
+  return numericProps['cartodb_id'] || numericProps['geoid'] || properties['cartodb_id'] || properties['geoid'] || (map.size + 1);
+}
+
+function getFeatureByType(coordinates, type) {
+  switch (type) {
+    case GEOMETRY_TYPES['Polygon']:
+      return { type: 'Polygon', coordinates: [coordinates] };
+    case GEOMETRY_TYPES['LineString']:
+      return { type: 'LineString', coordinates };
+    case GEOMETRY_TYPES['Point']:
+      return { type: 'Point', coordinates: coordinates[0] };
+    default:
+      throw new Error('Invalid geometry type');
+  }
+}
+
+function getRingCoordinatesFor(startIndex, endIndex, positions) {
+  const ringCoordinates = [];
+
+  for (let j = startIndex; j < endIndex; j++) {
+    ringCoordinates.push(
+      Array.from(positions.value.subarray(j * positions.size, (j + 1) * positions.size))
+    );
+  }
+
+  return ringCoordinates;
+}
+
+function calculateViewportFeatures({
+  map,
+  tileIsFullyVisible,
+  viewportIntersection,
+  data,
+  type,
+  uniqueIdProperty
+}) {
+  if (tileIsFullyVisible) {
+    // All the features of the tile are visible
+    const indices = getIndices(data);
+
+    for (let i = 0; i < indices.length - 1; i++) {
```

Comment:
@padawannn Is this < (length - 1) right? will we miss the last index here?

Reply:
Yes, it's right

Reply:
check this line https://github.com/CartoDB/carto-react/blob/25c83daa6803a8573a1242dfe6e15597239ce799/packages/react-core/src/filters/viewportFeaturesBinary.js#L17

Reply:
Oh, I see. Then it looks 👍🏻